--- libdispatch-1.3.orig/tests/bsdtests.c	2016-10-22 01:13:49.254193115 +0300
+++ libdispatch-1.3/tests/bsdtests.c	2016-10-12 01:16:52.000000000 +0300
@@ -25,7 +25,13 @@
 #include <stdarg.h>
 #include <stdio.h>
 #include <stdlib.h>
-#include <unistd.h>
+// Fix RedHat unistd.h
+#pragma push_macro("__block")
+#undef __block
+#define __block my__block
+#include_next <unistd.h>
+#pragma pop_macro("__block")
+//#include <unistd.h>
 #include <errno.h>
 #include <sys/errno.h>
 #include <sys/wait.h>
