diff -Nru gnustep-gui-0.25.0.orig/Model/GNUmakefile gnustep-gui-0.25.0/Model/GNUmakefile
--- gnustep-gui-0.25.0.orig/Model/GNUmakefile	2016-06-16 23:09:00.000000000 +0300
+++ gnustep-gui-0.25.0/Model/GNUmakefile	2019-03-29 14:14:03.505665001 +0200
@@ -66,9 +66,9 @@
 ADDITIONAL_LIB_DIRS += -L$(GNUSTEP_OBJ_DIR) 
 ifeq ($(GUI_LIB), gnu)
   ADDITIONAL_LIB_DIRS += -L../Source/$(GNUSTEP_OBJ_DIR)
-  ADDITIONAL_INCLUDE_DIRS = -I../Headers/Additions -I../Headers
+  ADDITIONAL_INCLUDE_DIRS += -I../Headers/Additions -I../Headers
 else
-  ADDITIONAL_INCLUDE_DIRS = -I../Headers/Additions
+  ADDITIONAL_INCLUDE_DIRS += -I../Headers/Additions
 endif
 
 ADDITIONAL_OBJC_FLAGS += $(BACKEND_DEFINE)
