--- gnustep-gui-0.24.1.orig/Source/NSApplication.m	2015-03-27 23:33:24.000000000 +0200
+++ gnustep-gui-0.24.1/Source/NSApplication.m	2016-11-13 04:13:22.654949748 +0200
@@ -1334,6 +1335,13 @@
           postNotificationName: NSApplicationDidBecomeActiveNotification
 		      object: [NSWorkspace sharedWorkspace]
           userInfo: info];
+
+      if ([[[NSUserDefaults standardUserDefaults]
+             objectForKey: @"NXAutoLaunch"] isEqualToString: @"YES"])
+        {
+          _app_is_active = NO;
+          [[NSUserDefaults standardUserDefaults] removeSuiteNamed: NSArgumentDomain];
+        }
     }
 }
 
