--- gnustep-back-0.24.1.orig/Source/x11/XGServer.m	2012-01-18 23:35:59.000000000 +0200
+++ gnustep-back-0.24.1/Source/x11/XGServer.m	2017-12-15 18:18:15.757639843 +0200
@@ -40,6 +40,7 @@
 #include <Foundation/NSString.h>
 #include <Foundation/NSUserDefaults.h>
 #include <Foundation/NSDebug.h>
+#include <Foundation/NSDistributedNotificationCenter.h>
 
 #include <signal.h>
 /* Terminate cleanly if we get a signal to do so */
@@ -487,6 +488,7 @@
 - (void) dealloc
 {
   NSDebugLog(@"Destroying X11 Server");
+  [[NSDistributedNotificationCenter defaultCenter] removeObserver:self];
   DESTROY(inputServer);
   [self _destroyServerWindows];
   NSFreeMapTable(screenList);
