--- gnustep-gui-0.24.1.orig/Source/NSWindow.m	2016-10-27 22:04:22.053402864 +0300
+++ gnustep-gui-0.24.1/Source/NSWindow.m	2016-10-31 13:28:14.280311891 +0200
@@ -479,7 +479,8 @@
   
   iconSize = GSGetIconSize();
   
-  tileImage = [[GSCurrentServer() iconTileImage] copy];
+  // tileImage = [[GSCurrentServer() iconTileImage] copy];
+  tileImage = [[NSImage imageNamed:@"common_MiniWindowTile"] copy];
   [tileImage setScalesWhenResized: YES];
   [tileImage setSize: iconSize];
   
@@ -512,8 +513,8 @@
                                  iconSize.width - ((iconSize.width / 8) * 2),
                                  iconSize.height - ((iconSize.height / 8) * 2))
               inView: self];
-  [titleCell drawWithFrame: NSMakeRect(1, iconSize.height - 12,
-                                       iconSize.width - 2, 11)
+  [titleCell drawWithFrame: NSMakeRect(2, iconSize.height - 12,
+                                       iconSize.width - 3, 10)
                     inView: self];
 }
 
@@ -603,7 +604,7 @@
       [titleCell setDrawsBackground: YES];
       [titleCell setBackgroundColor: [NSColor blackColor]];
       [titleCell setTextColor: [NSColor whiteColor]];
-      [titleCell setFont: [NSFont systemFontOfSize: 8]];
+      [titleCell setFont: [NSFont systemFontOfSize: 9]];
     }
   else
     {
