--- gnustep-gui-0.24.1.orig/Source/NSApplication.m	2015-03-27 23:33:24.000000000 +0200
+++ gnustep-gui-0.24.1/Source/NSApplication.m	2018-12-22 19:25:36.952061108 +0200
@@ -433,6 +433,11 @@
   return NO;
 }
 
+- (BOOL) becomesKeyOnlyIfNeeded
+{
+  return YES;
+}
+
 - (BOOL) worksWhenModal
 {
   return YES;
@@ -541,6 +546,11 @@
   return YES;
 }
 
+- (BOOL) needsPanelToBecomeKey
+{
+  return NO;
+}
+
 - (void) concludeDragOperation: (id<NSDraggingInfo>)sender
 {
 }
@@ -1334,6 +1344,13 @@
           postNotificationName: NSApplicationDidBecomeActiveNotification
 		      object: [NSWorkspace sharedWorkspace]
           userInfo: info];
+
+      if ([[[NSUserDefaults standardUserDefaults]
+             objectForKey: @"NXAutoLaunch"] isEqualToString: @"YES"])
+        {
+          _app_is_active = NO;
+          [[NSUserDefaults standardUserDefaults] removeSuiteNamed: NSArgumentDomain];
+        }
     }
 }
 
@@ -2431,71 +2448,74 @@
 	{
 	  NSArray		*windows_list;
 	  NSDictionary  	*info;
-	  NSWindow		*win;
+	  NSWindow		*win = [[self mainMenu] window];
 	  NSEnumerator  	*iter;
 
 	  [nc postNotificationName: NSApplicationWillHideNotification
 	                    object: self];
 
-	  if ([self keyWindow] != nil)
-	    {
-	      _hidden_key = [self keyWindow];
-	      [_hidden_key resignKeyWindow];
-	    }
+          if ([GSServerForWindow(win) hidewindow: [win windowNumber]] == NO)
+            {
+              if ([self keyWindow] != nil)
+                {
+                  _hidden_key = [self keyWindow];
+                  [_hidden_key resignKeyWindow];
+                }
 	  
-	  // The main window is saved for when the app is activated again.
-	  // This is necessary for menu in window.
-	  if ([self mainWindow] != nil)
-	    {
-	      _hidden_main = [self mainWindow];
-	      [_hidden_main resignMainWindow];
-	    }
+              // The main window is saved for when the app is activated again.
+              // This is necessary for menu in window.
+              if ([self mainWindow] != nil)
+                {
+                  _hidden_main = [self mainWindow];
+                  [_hidden_main resignMainWindow];
+                }
 	  
-	  windows_list = GSOrderedWindows();
-	  iter = [windows_list reverseObjectEnumerator];
+              windows_list = GSOrderedWindows();
+              iter = [windows_list reverseObjectEnumerator];
 	  
-	  while ((win = [iter nextObject]))
-	    {
-	      if ([win isVisible] == NO && ![win isMiniaturized])
-		{
-		  continue;		/* Already invisible	*/
-		}
-	      if ([win canHide] == NO)
-		{
-		  continue;		/* Not hideable	*/
-		}
-	      if (win == _app_icon_window)
-		{
-		  continue;		/* can't hide the app icon.	*/
-		}
-	      if (_app_is_active == YES && [win hidesOnDeactivate] == YES)
-		{
-		  continue;		/* Will be hidden by deactivation	*/
-		}
-	      [_hidden addObject: win];
-	      [win orderOut: self];
-	    }
-	  _app_is_hidden = YES;
+              while ((win = [iter nextObject]))
+                {
+                  if ([win isVisible] == NO && ![win isMiniaturized])
+                    {
+                      continue;		/* Already invisible	*/
+                    }
+                  if ([win canHide] == NO)
+                    {
+                      continue;		/* Not hideable	*/
+                    }
+                  if (win == _app_icon_window)
+                    {
+                      continue;		/* can't hide the app icon.	*/
+                    }
+                  if (_app_is_active == YES && [win hidesOnDeactivate] == YES)
+                    {
+                      continue;		/* Will be hidden by deactivation	*/
+                    }
+                  [_hidden addObject: win];
+                  [win orderOut: self];
+                }
+            }
+          _app_is_hidden = YES;
 	  
-	  if (YES == [[NSUserDefaults standardUserDefaults]
-		       boolForKey: @"GSSuppressAppIcon"])
-	    {
+          if (YES == [[NSUserDefaults standardUserDefaults]
+                           boolForKey: @"GSSuppressAppIcon"])
+            {
 #if	MINI_ICON
-	      NSRect	f = [[[self mainMenu] window] frame];
-	      NSPoint	p = f.origin;
+              NSRect	f = [[[self mainMenu] window] frame];
+              NSPoint	p = f.origin;
 	      
-	      p.y += f.size.height;
-	      [_app_icon_window setFrameTopLeftPoint: p];
-	      [_app_icon_window orderFrontRegardless];
-	      [_app_icon_window miniaturize: self];
+              p.y += f.size.height;
+              [_app_icon_window setFrameTopLeftPoint: p];
+              [_app_icon_window orderFrontRegardless];
+              [_app_icon_window miniaturize: self];
 #else
-	      [_app_icon_window orderFrontRegardless];
+              [_app_icon_window orderFrontRegardless];
 #endif
-	    }
-	  else
-	    {
-	      [[_app_icon_window contentView] setNeedsDisplay: YES];
-	    }
+            }
+          else
+            {
+              [[_app_icon_window contentView] setNeedsDisplay: YES];
+            }
 	  
 	  /*
 	   * On hiding we also deactivate the application which will make the menus
