--- gnustep-back-0.24.1.orig/Headers/x11/XGServerWindow.h	2011-06-01 23:56:27.000000000 +0300
+++ gnustep-back-0.24.1/Headers/x11/XGServerWindow.h	2018-07-13 02:26:38.660067069 +0300
@@ -69,10 +69,8 @@
 #define GSWindowWillResizeNotificationsFlag	(1<<1)
 #define GSWindowWillMoveNotificationsFlag 	(1<<2)
 #define GSNoApplicationIconFlag			(1<<5)
-#define WMFHideOtherApplications		10
-#define WMFHideApplication			12
 
-#define GSMaxWMProtocols 5
+#define GSMaxWMProtocols 6
 
 /* Graphics Driver protocol. Setup in [NSGraphicsContext-contextDevice:] */
 enum {
