diff -ruN a/Applications/Gorm/GNUmakefile b/Applications/Gorm/GNUmakefile
--- a/Applications/Gorm/GNUmakefile	2024-05-30 00:10:32.000000000 -0500
+++ b/Applications/Gorm/GNUmakefile	2024-05-30 22:50:28.932738959 -0500
@@ -148,6 +148,7 @@
 Gorm_LOCALIZED_RESOURCE_FILES = \
 	Gorm.gorm \
 	GormLanguageViewController.gorm \
+	InfoPanel.gorm \
 	Gorm.rtfd
 
 Gorm_LANGUAGES = \
diff -ruN a/GormCore/GormDocumentController.h b/GormCore/GormDocumentController.h
--- a/GormCore/GormDocumentController.h	2024-05-30 00:10:32.000000000 -0500
+++ b/GormCore/GormDocumentController.h	2024-05-30 22:42:53.829256137 -0500
@@ -41,10 +41,16 @@
 
 @interface GormDocumentController : NSDocumentController
 {
+  id infoPanel;
+  id infoWindow;
+  IBOutlet NSTextField* versionField;
+  IBOutlet NSTextField* copyrightField;
+  NSDictionary *infoDict;
 }
 
 - (void) buildDocumentForType: (GormDocumentType)documentType;
 - (id) openDocumentWithContentsOfURL:(NSURL *)url;
+- (void) showInfoPanel:(id)sender;
 
 @end
 
diff -ruN a/GormCore/GormDocumentController.m b/GormCore/GormDocumentController.m
--- a/GormCore/GormDocumentController.m	2024-05-30 00:10:32.000000000 -0500
+++ b/GormCore/GormDocumentController.m	2024-05-30 22:45:42.331108944 -0500
@@ -31,6 +31,30 @@
 
 @implementation GormDocumentController
 
+- (id)init
+{
+  if ((self = [super init]))
+    {
+      NSString *file;
+
+      file = [[NSBundle mainBundle] pathForResource:@"GormInfo"
+        ofType:@"plist"];
+
+      infoDict = [NSDictionary dictionaryWithContentsOfFile:file];
+      RETAIN(infoDict);
+
+      if ([NSBundle loadNibNamed:@"Info" owner:self] == NO)
+        {
+          return nil;
+        }
+      [versionField setStringValue:[NSString stringWithFormat:@"Version %@", [infoDict objectForKey:@"ApplicationRelease"]]];
+      [copyrightField setStringValue:[infoDict objectForKey:@"Copyright"]];
+      [infoWindow center];
+    }
+
+  return self;
+}
+
 - (id) currentDocument
 {
   NSArray  *documents = [self documents];
@@ -206,4 +230,17 @@
 {
   return  [self openDocumentWithContentsOfURL:url display:YES];
 }
+
+- (void)showInfoPanel:(id)sender
+{
+  if (!infoPanel) {
+     if (![NSBundle loadNibNamed:@"InfoPanel" owner:self]) {
+        NSLog (@"Failed to load InfoPanel.gorm");
+        return;
+      }
+     [infoPanel center];
+   }
+ [infoPanel makeKeyAndOrderFront:nil];
+}
+
 @end
