diff -ruN a/GNUmakefile b/GNUmakefile
--- a/GNUmakefile	2022-10-28 12:41:54.000000000 -0400
+++ b/GNUmakefile	2023-08-05 23:28:16.539774998 -0400
@@ -181,6 +181,7 @@
 	GormShelfPref.gorm \
 	GormSoundInspector.gorm \
 	GormViewSizeInspector.gorm \
+	InfoPanel.gorm \
 	Gorm.rtfd
 
 Gorm_LANGUAGES = \
diff -ruN a/GormCore/GormDocumentController.h b/GormCore/GormDocumentController.h
--- a/GormCore/GormDocumentController.h	2022-10-28 12:41:54.000000000 -0400
+++ b/GormCore/GormDocumentController.h	2023-08-06 00:10:48.308297018 -0400
@@ -41,10 +41,16 @@
 
 @interface GormDocumentController : NSDocumentController
 {
+  id infoPanel;
+  id infoWindow;
+  IBOutlet NSTextField* versionField;
+  IBOutlet NSTextField* copyrightField;
+  NSDictionary *infoDict;
 }
 
 - (void) buildDocumentForType: (GormDocumentType)documentType;
 - (id) openDocumentWithContentsOfURL:(NSURL *)url;
+- (void) showInfoPanel:(id)sender;
 
 @end
 
diff -ruN a/GormCore/GormDocumentController.m b/GormCore/GormDocumentController.m
--- a/GormCore/GormDocumentController.m	2022-10-28 12:41:54.000000000 -0400
+++ b/GormCore/GormDocumentController.m	2023-08-05 23:32:23.339052057 -0400
@@ -30,6 +30,30 @@
 
 @implementation GormDocumentController
 
+- (id)init
+{
+  if ((self = [super init]))
+    {
+      NSString *file;
+
+      file = [[NSBundle mainBundle] pathForResource:@"GormInfo"
+        ofType:@"plist"];
+
+      infoDict = [NSDictionary dictionaryWithContentsOfFile:file];
+      RETAIN(infoDict);
+
+      if ([NSBundle loadNibNamed:@"Info" owner:self] == NO)
+        {
+          return nil;
+        }
+      [versionField setStringValue:[NSString stringWithFormat:@"Version %@", [infoDict objectForKey:@"ApplicationRelease"]]];
+      [copyrightField setStringValue:[infoDict objectForKey:@"Copyright"]];
+      [infoWindow center];
+    }
+
+  return self;
+}
+
 - (id) currentDocument
 {
   NSArray  *documents = [self documents];
@@ -205,4 +229,17 @@
 {
   return  [self openDocumentWithContentsOfURL:url display:YES];
 }
+
+- (void)showInfoPanel:(id)sender
+{
+  if (!infoPanel) {
+     if (![NSBundle loadNibNamed:@"InfoPanel" owner:self]) {
+        NSLog (@"Failed to load InfoPanel.gorm");
+        return;
+      }
+     [infoPanel center];
+   }
+ [infoPanel makeKeyAndOrderFront:nil];
+}
+
 @end
