diff -ruN a/Source/NSDocument.m b/Source/NSDocument.m
--- a/Source/NSDocument.m	2022-12-29 04:05:31.000000000 -0500
+++ b/Source/NSDocument.m	2023-10-02 22:49:20.730125876 -0500
@@ -1138,8 +1138,8 @@
 {
   if (_save_panel_accessory == nil)
     {
-      NSRect accessoryFrame = NSMakeRect(0,0,380,70);
-      NSRect spaFrame = NSMakeRect(115,14,150,22);
+      NSRect accessoryFrame = NSMakeRect(0,0,292,60);
+      NSRect spaFrame = NSMakeRect(63,6,150,22);
 
       _save_panel_accessory = [[NSBox alloc] initWithFrame: accessoryFrame];
       [(NSBox *)_save_panel_accessory setTitle: _(@"File Type")];
