--- a/Headers/Additions/GNUstepGUI/GSDisplayServer.h	2022-12-29 04:05:31.000000000 -0500
+++ b/Headers/Additions/GNUstepGUI/GSDisplayServer.h	2023-10-02 23:13:30.863155782 -0500
@@ -117,6 +117,14 @@
 /* Screen capture */ 
 - (NSImage *) contentsOfScreen: (int)screen inRect: (NSRect)rect;
 
+/* Desktops (workspaces) */
+- (unsigned int) numberOfDesktops: (int)screen;
+- (NSArray *) namesOfDesktops: (int)screen;
+- (unsigned int) desktopNumberForScreen: (int)screen;
+- (void) setDesktopNumber: (unsigned int)workspace forScreen: (int)screen;
+- (unsigned int) desktopNumberForWindow: (int)win;
+- (void) setDesktopNumber: (unsigned int)workspace forWindow: (int)win;
+
 @end
 
 /* ----------------------------------------------------------------------- */
--- a/Source/GSDisplayServer.m	2022-12-29 04:05:31.000000000 -0500
+++ b/Source/GSDisplayServer.m	2023-10-03 21:54:36.242319328 -0500
@@ -510,6 +510,40 @@
   return nil;
 }
 
+- (unsigned int) numberOfDesktops: (int)screen
+{
+  [self subclassResponsibility: _cmd];
+  return 1;
+}
+
+- (NSArray *) namesOfDesktops: (int)screen
+{
+  [self subclassResponsibility: _cmd];
+  return nil;
+}
+
+- (unsigned int) desktopNumberForScreen: (int)screen
+{
+  [self subclassResponsibility: _cmd];
+  return 0;
+}
+
+- (void) setDesktopNumber: (unsigned int)workspace forScreen: (int)screen
+{
+  // Do nothing if not overridden by subclass
+}
+
+- (unsigned int) desktopNumberForWindow: (int)win
+{
+  [self subclassResponsibility: _cmd];
+  return 0;
+}
+
+- (void) setDesktopNumber: (unsigned int)workspace forWindow: (int)win
+{
+  // Do nothing if not overridden by subclass
+}
+
 @end
 
 /* ----------------------------------------------------------------------- */
