--- gnustep-back-0.28.0.orig/Source/x11/XGServerWindow.m	2019-12-16 12:02:02.128214723 +0200
+++ gnustep-back-0.28.0/Source/x11/XGServerWindow.m	2019-12-16 12:16:11.169214723 +0200
@@ -2018,8 +2014,7 @@
 
   // For window managers supporting EWMH, but not Window Maker,
   // where we use a different solution, set the window icon.
-  if (((generic.wm & XGWM_EWMH) != 0)
-    && ((generic.wm & XGWM_WINDOWMAKER) == 0))
+  if ((generic.wm & XGWM_EWMH) != 0)
     {
       [self _setNetWMIconFor: window->ident];
     }
