diff -ruN a/Source/NSPrintPanel.m b/Source/NSPrintPanel.m
--- a/Source/NSPrintPanel.m	2022-12-29 04:05:31.000000000 -0500
+++ b/Source/NSPrintPanel.m	2023-10-02 22:53:18.010041994 -0500
@@ -116,7 +116,7 @@
 //
 - (id) init
 {
-  return [self initWithContentRect: NSMakeRect(300, 300, 420, 350)
+  return [self initWithContentRect: NSMakeRect(300, 300, 420, 300)
 			 styleMask: NSTitledWindowMask
 			   backing: NSBackingStoreBuffered
 			     defer: YES];
@@ -343,8 +343,8 @@
 
 - (NSBox *) _savePanelAccessory
 {
-  NSRect accessoryFrame = NSMakeRect(0,0,380,70);
-  NSRect spaFrame = NSMakeRect(115,14,150,22);
+  NSRect accessoryFrame = NSMakeRect(0,0,292,60);
+  NSRect spaFrame = NSMakeRect(60,6,150,22);
   
   NSBox *save_panel_accessory = [[[NSBox alloc] initWithFrame: accessoryFrame] autorelease];
   [save_panel_accessory setTitle: _(@"File Type")];
