diff -ruN a/Source/NSSavePanel.m b/Source/NSSavePanel.m
--- a/Source/NSSavePanel.m	2022-12-29 04:05:31.000000000 -0500
+++ b/Source/NSSavePanel.m	2023-10-02 22:57:29.211920051 -0500
@@ -235,7 +235,7 @@
   [_browser setAction: @selector(_selectText:)];
   [_browser setDoubleAction: @selector(_doubleClick:)];
   [_browser setTarget: self];
-  [_browser setMinColumnWidth: 140];
+  [_browser setMinColumnWidth: 120];
   [_topView addSubview: _browser];
   [_browser release];
   
@@ -377,7 +377,7 @@
   [_topView addSubview: bar];
   [bar release];
 
-  [self setContentSize: NSMakeSize (384, 426)];
+  [self setContentSize: NSMakeSize (308, 317)];
   [self setInitialFirstResponder: _form];
   [super setTitle: @""];
 
