--- gnustep-gui-0.24.1.orig/Source/NSWindow.m	2014-12-26 09:59:29.000000000 +0200
+++ gnustep-gui-0.24.1/Source/NSWindow.m	2018-07-13 01:21:49.555146429 +0300
@@ -479,7 +479,7 @@
   
   iconSize = GSGetIconSize();
   
-  tileImage = [[GSCurrentServer() iconTileImage] copy];
+  tileImage = [[NSImage imageNamed:@"common_MiniWindowTile"] copy];
   [tileImage setScalesWhenResized: YES];
   [tileImage setSize: iconSize];
   
@@ -512,8 +512,8 @@
                                  iconSize.width - ((iconSize.width / 8) * 2),
                                  iconSize.height - ((iconSize.height / 8) * 2))
               inView: self];
-  [titleCell drawWithFrame: NSMakeRect(1, iconSize.height - 12,
-                                       iconSize.width - 2, 11)
+  [titleCell drawWithFrame: NSMakeRect(2, iconSize.height - 12,
+                                       iconSize.width - 3, 10)
                     inView: self];
 }
 
@@ -603,7 +603,7 @@
       [titleCell setDrawsBackground: YES];
       [titleCell setBackgroundColor: [NSColor blackColor]];
       [titleCell setTextColor: [NSColor whiteColor]];
-      [titleCell setFont: [NSFont systemFontOfSize: 8]];
+      [titleCell setFont: [NSFont systemFontOfSize: 9]];
     }
   else
     {
@@ -4195,6 +4195,10 @@
               [self performMiniaturize: NSApp];
               break;
 
+            case GSAppKitAppHide:
+              [NSApp hide: self];
+              break;
+
             case GSAppKitWindowFocusIn:
               if (_f.is_miniaturized)
 		{
@@ -4212,8 +4216,8 @@
               if ([self canBecomeKeyWindow] == YES)
 		{
 		  NSDebugLLog(@"Focus", @"Making %d key", (int)_windowNum);
-		  [self makeKeyWindow];
-		  [self makeMainWindow];
+                  [self makeKeyWindow];
+                  [self makeMainWindow];
 		  [NSApp activateIgnoringOtherApps: YES];
 		}
               if (self == [[NSApp mainMenu] window])
