diff -Nru a/CoreFoundation/Base.subproj/CFInternal.h b/CoreFoundation/Base.subproj/CFInternal.h
--- a/CoreFoundation/Base.subproj/CFInternal.h	2020-11-20 09:31:35.000000000 +0200
+++ b/CoreFoundation/Base.subproj/CFInternal.h	2020-11-23 12:39:49.000000000 +0200
@@ -116,7 +116,12 @@
 #endif // TARGET_OS_MAC || TARGET_OS_LINUX || TARGET_OS_BSD || TARGET_OS_WASI
 
 #if __has_include(<unistd.h>)
-#include <unistd.h>
+/* Fix CentOS 7 problematic unistd.h */
+#pragma push_macro("__block")
+  #undef __block
+  #define __block my__block
+  #include <unistd.h>
+#pragma pop_macro("__block")
 #endif
 #if _POSIX_THREADS
 #include <pthread.h>
diff -Nru a/CoreFoundation/Base.subproj/CFPriv.h b/CoreFoundation/Base.subproj/CFPriv.h
--- a/CoreFoundation/Base.subproj/CFPriv.h	2020-11-23 12:44:53.962386591 +0200
+++ b/CoreFoundation/Base.subproj/CFPriv.h	2020-11-23 12:40:41.000000000 +0200
@@ -585,7 +585,12 @@
 #endif
 
 #if __has_include(<unistd.h>)
-#include <unistd.h>
+/* Fix CentOS 7 problematic unistd.h */
+#pragma push_macro("__block")
+  #undef __block
+  #define __block my__block
+  #include <unistd.h>
+#pragma pop_macro("__block")
 #endif
 #if _POSIX_THREADS
 #include <pthread.h>
diff -Nru a/CoreFoundation/Locale.subproj/CFICULogging.h b/CoreFoundation/Locale.subproj/CFICULogging.h
--- a/CoreFoundation/Locale.subproj/CFICULogging.h	2020-11-20 09:31:35.000000000 +0200
+++ b/CoreFoundation/Locale.subproj/CFICULogging.h	2020-11-23 12:46:11.082927573 +0200
@@ -19,7 +19,9 @@
 #include <unicode/udatpg.h>
 #include <unicode/udat.h>
 #include <unicode/unum.h>
+#if __has_include(<unicode/ulistformatter.h>)
 #include <unicode/ulistformatter.h>
+#endif
 #include <unicode/ucurr.h>
 #include <unicode/ustring.h>
 #if __has_include(<unicode/ureldatefmt.h>)
diff -Nru a/CoreFoundation/Locale.subproj/CFLocale.c b/CoreFoundation/Locale.subproj/CFLocale.c
--- a/CoreFoundation/Locale.subproj/CFLocale.c	2020-11-23 12:44:53.979395090 +0200
+++ b/CoreFoundation/Locale.subproj/CFLocale.c	2020-11-23 12:46:11.084928573 +0200
@@ -32,7 +32,9 @@
 #include <unicode/putil.h>          // ICU low-level utilities
 #include <unicode/umsg.h>           // ICU message formatting
 #include <unicode/ucol.h>
+#if __has_include(<unicode/unumsys.h>)
 #include <unicode/unumsys.h>        // ICU numbering systems
+#endif
 #include <unicode/uvernum.h>
 #if U_ICU_VERSION_MAJOR_NUM > 53 && __has_include(<unicode/uameasureformat.h>)
 #include <unicode/uameasureformat.h>
@@ -192,7 +194,7 @@
 }
 
 CF_INLINE Boolean __CFLocaleCacheGetValueIfPresent(CFLocaleRef locale, CFLocaleKey key, CFTypeRef *value) {
-    CFDictionaryRef cache = atomic_load_explicit(&locale->_cache, memory_order_acquire);
+    CFDictionaryRef cache = atomic_load_explicit((_Atomic(CFMutableDictionaryRef) *)&locale->_cache, memory_order_acquire);
     if (!cache) {
         *value = NULL;
         return false;
@@ -202,7 +204,7 @@
 }
 
 CF_INLINE void __CFLocaleCacheSet_alreadyLocked(CFLocaleRef locale, CFLocaleKey key, CFTypeRef value) {
-    CFMutableDictionaryRef cache = atomic_load_explicit(&locale->_cache, memory_order_acquire);
+    CFMutableDictionaryRef cache = atomic_load_explicit((_Atomic(CFMutableDictionaryRef) *)&locale->_cache, memory_order_acquire);
     if (!cache) {
         cache = CFDictionaryCreateMutable(CFGetAllocator(locale), 0, NULL, &kCFTypeDictionaryValueCallBacks);
         atomic_store_explicit(&((struct __CFLocale *)locale)->_cache, cache, memory_order_release);
diff -Nru a/CoreFoundation/RunLoop.subproj/CFSocket.c b/CoreFoundation/RunLoop.subproj/CFSocket.c
--- a/CoreFoundation/RunLoop.subproj/CFSocket.c	2020-11-20 09:31:35.000000000 +0200
+++ b/CoreFoundation/RunLoop.subproj/CFSocket.c	2020-11-23 12:42:15.000000000 +0200
@@ -34,7 +34,12 @@
 #include <sys/ioctl.h>
 #endif
 #if defined(__unix__) || (defined(__APPLE__) && defined(__MACH__))
-#include <unistd.h>
+/* Fix CentOS 7 problematic unistd.h */
+#pragma push_macro("__block")
+  #undef __block
+  #define __block my__block
+  #include <unistd.h>
+#pragma pop_macro("__block")
 #endif
 #include <fcntl.h>
 #include <CoreFoundation/CFArray.h>
diff -Nru a/CoreFoundation/String.subproj/CFRegularExpression.c b/CoreFoundation/String.subproj/CFRegularExpression.c
--- a/CoreFoundation/String.subproj/CFRegularExpression.c	2020-11-20 09:31:35.000000000 +0200
+++ b/CoreFoundation/String.subproj/CFRegularExpression.c	2020-11-23 12:46:11.084928573 +0200
@@ -171,41 +171,6 @@
     return (CFIndex)uregex_groupCount(regex->regex, &errorCode);
 }
 
-CFIndex _CFRegularExpressionGetCaptureGroupNumberWithName(_CFRegularExpressionRef regex, CFStringRef groupName) {
-    UniChar stackBuffer[STACK_BUFFER_SIZE], *nameBuffer = NULL;
-    Boolean freeNameBuffer = false;
-
-    CFIndex nameLength = CFStringGetLength(groupName);
-    UErrorCode errorCode = U_ZERO_ERROR;
-
-    nameBuffer = (UniChar *)CFStringGetCharactersPtr(groupName);
-    if (!nameBuffer) {
-        if (nameLength <= STACK_BUFFER_SIZE) {
-            nameBuffer = stackBuffer;
-            CFStringGetCharacters(groupName, CFRangeMake(0, nameLength), nameBuffer);
-        } else {
-            nameBuffer = (UniChar *)malloc(sizeof(UniChar) * nameLength);
-            if (nameBuffer) {
-                CFStringGetCharacters(groupName, CFRangeMake(0, nameLength), nameBuffer);
-                freeNameBuffer = true;
-            } else {
-                HALT;
-            }
-        }
-    }
-
-    CFIndex idx = uregex_groupNumberFromName(regex->regex, nameBuffer, nameLength, &errorCode);
-    if (U_FAILURE(errorCode) || idx < 0) {
-        idx = kCFNotFound;
-    }
-
-    if (freeNameBuffer) {
-        free(nameBuffer);
-    }
-
-    return idx;
-}
-
 struct regexCallBackContext {
     void *context;
     void (*match)(void *context, CFRange *ranges, CFIndex count, _CFRegularExpressionMatchingFlags flags, Boolean *stop);
diff -Nru a/CoreFoundation/String.subproj/CFRegularExpression.h b/CoreFoundation/String.subproj/CFRegularExpression.h
--- a/CoreFoundation/String.subproj/CFRegularExpression.h	2020-11-20 09:31:35.000000000 +0200
+++ b/CoreFoundation/String.subproj/CFRegularExpression.h	2020-11-23 12:46:11.085929074 +0200
@@ -57,7 +57,6 @@
 void _CFRegularExpressionDestroy(_CFRegularExpressionRef regex);
 
 CFIndex _CFRegularExpressionGetNumberOfCaptureGroups(_CFRegularExpressionRef regex);
-CFIndex _CFRegularExpressionGetCaptureGroupNumberWithName(_CFRegularExpressionRef regex, CFStringRef groupName);
 void _CFRegularExpressionEnumerateMatchesInString(_CFRegularExpressionRef regexObj, CFStringRef string, _CFRegularExpressionMatchingOptions options, CFRange range, void *_Nullable context, _CFRegularExpressionMatch match);
 
 CFStringRef _CFRegularExpressionGetPattern(_CFRegularExpressionRef regex);
diff -Nru a/CoreFoundation/String.subproj/CFString.c b/CoreFoundation/String.subproj/CFString.c
--- a/CoreFoundation/String.subproj/CFString.c	2020-11-23 12:44:53.985398090 +0200
+++ b/CoreFoundation/String.subproj/CFString.c	2020-11-23 12:46:11.088930574 +0200
@@ -3550,7 +3550,7 @@
     //   Grapheme_Extend = Yes, or
     //   Emoji_Modifier=Yes in emoji-data.txt
     //
-    return u_hasBinaryProperty(character, UCHAR_GRAPHEME_EXTEND) || u_hasBinaryProperty(character, UCHAR_EMOJI_MODIFIER);
+    return u_hasBinaryProperty(character, UCHAR_GRAPHEME_EXTEND);
 }
 
 static inline bool __CFStringIsValidExtendedPictographicCharacterForPictographicSequence(UTF32Char character) {
