diff --git a/CoreFoundation/Base.subproj/CFRuntime.c b/CoreFoundation/Base.subproj/CFRuntime.c
index fde9ef10..19167194 100644
--- a/CoreFoundation/Base.subproj/CFRuntime.c
+++ b/CoreFoundation/Base.subproj/CFRuntime.c
@@ -241,6 +241,7 @@ CFRuntimeClass const * __CFRuntimeClassTable[__CFRuntimeClassTableSize * 2] __at
     [_kCFRuntimeIDCFRunLoopSource] = &__CFRunLoopSourceClass,
     [_kCFRuntimeIDCFRunLoopObserver] = &__CFRunLoopObserverClass,
     [_kCFRuntimeIDCFRunLoopTimer] = &__CFRunLoopTimerClass,
+    [_kCFRuntimeIDCFFileDescriptor] = &__CFFileDescriptorClass,
     [_kCFRuntimeIDCFSocket] = &__CFSocketClass,
     [_kCFRuntimeIDCFReadStream] = &__CFReadStreamClass,
     [_kCFRuntimeIDCFWriteStream] = &__CFWriteStreamClass,
diff --git a/CoreFoundation/Base.subproj/CFRuntime_Internal.h b/CoreFoundation/Base.subproj/CFRuntime_Internal.h
index edfe7b9f..9a0345ff 100644
--- a/CoreFoundation/Base.subproj/CFRuntime_Internal.h
+++ b/CoreFoundation/Base.subproj/CFRuntime_Internal.h
@@ -133,6 +133,7 @@ CF_PRIVATE const CFRuntimeClass __CFRunLoopClass;
 CF_PRIVATE const CFRuntimeClass __CFRunLoopSourceClass;
 CF_PRIVATE const CFRuntimeClass __CFRunLoopObserverClass;
 CF_PRIVATE const CFRuntimeClass __CFRunLoopTimerClass;
+CF_PRIVATE const CFRuntimeClass __CFFileDescriptorClass;
 CF_PRIVATE const CFRuntimeClass __CFSocketClass;
 CF_PRIVATE const CFRuntimeClass __CFReadStreamClass;
 CF_PRIVATE const CFRuntimeClass __CFWriteStreamClass;
diff --git a/CoreFoundation/CMakeLists.txt b/CoreFoundation/CMakeLists.txt
index c146246a..63dafee9 100644
--- a/CoreFoundation/CMakeLists.txt
+++ b/CoreFoundation/CMakeLists.txt
@@ -235,6 +235,7 @@ add_framework(CoreFoundation
                 RunLoop.subproj/CFMessagePort.h
                 RunLoop.subproj/CFRunLoop.h
                 RunLoop.subproj/CFSocket.h
+                RunLoop.subproj/CFFileDescriptor.h
                 # Stream
                 Stream.subproj/CFStream.h
                 # String
@@ -317,6 +318,7 @@ add_framework(CoreFoundation
                 # RunLoop.subproj/CFMessagePort.c
                 RunLoop.subproj/CFRunLoop.c
                 RunLoop.subproj/CFSocket.c
+                RunLoop.subproj/CFFileDescriptor.c
                 # Stream
                 Stream.subproj/CFConcreteStreams.c
                 Stream.subproj/CFSocketStream.c
