--- gnustep-back-0.24.1.orig/Headers/x11/XGGeneric.h	2015-03-21 20:53:46.000000000 +0200
+++ gnustep-back-0.24.1/Headers/x11/XGGeneric.h	2018-07-13 00:57:48.244175840 +0300
@@ -116,6 +116,7 @@
   Atom                  net_wm_sync_request_atom;
   Atom                  net_wm_sync_request_counter_atom;
   Atom			miniaturize_atom;
+  Atom			hide_app_atom;
   Atom			win_decor_atom;
   Atom			titlebar_state_atom;
   char			*rootName;
