--- gnustep-gui-0.24.1.orig/Source/GSDragView.m	2013-02-18 22:29:10.000000000 +0200
+++ gnustep-gui-0.24.1/Source/GSDragView.m	2018-06-18 01:59:23.856328596 +0300
@@ -510,20 +510,26 @@
     {
     case NSDragOperationNone:
       name = @"NoCursor";
-      iname = @"common_noCursor";
+      iname = @"dragNoneCursor";
       break;
     case NSDragOperationCopy:
       name = @"CopyCursor";
-      iname = @"common_copyCursor";
+      iname = @"dragCopyCursor";
       break;
     case NSDragOperationLink:
       name = @"LinkCursor";
-      iname = @"common_linkCursor";
+      iname = @"dragLinkCursor";
       break;
-    case NSDragOperationGeneric:
+    case NSDragOperationDelete:
+      name = @"DeleteCursor";
+      iname = @"dragDeleteCursor";
       break;
     default:
-      // FIXME: Should not happen, add warning?
+      if (targetWindowRef != 0)
+        {
+          name = @"MoveCursor";
+          iname = @"dragMoveCursor";
+        }
       break;
     }
 
@@ -544,7 +550,6 @@
       newCursor = [cursors objectForKey: name];
       if (newCursor == nil)
 	{
-	  /* Make our own arrow cursor, since we want to color it */
 	  void *c;
 	  
 	  newCursor = [[NSCursor alloc] initWithImage: nil];
@@ -554,19 +559,6 @@
 	  RELEASE(newCursor);
 	}
     }
-  
-  if ((targetWindowRef != 0) && mask != NSDragOperationNone)
-    {
-      [GSCurrentServer() recolorcursor: [NSColor greenColor] 
-		      : [NSColor blackColor] 
-		      : [newCursor _cid]];
-    }
-  else
-    {
-      [GSCurrentServer() recolorcursor: [NSColor blackColor] 
-		      : [NSColor whiteColor] 
-		      : [newCursor _cid]];
-    }
 
   [newCursor set];
 }
