diff -Nru gnustep-gui-0.25.0.orig/Images/GNUmakefile gnustep-gui-0.25.0/Images/GNUmakefile
--- gnustep-gui-0.25.0.orig/Images/GNUmakefile	2016-06-16 23:09:00.000000000 +0300
+++ gnustep-gui-0.25.0/Images/GNUmakefile	2019-05-03 12:57:45.813818002 +0300
@@ -63,7 +63,6 @@
  common_CloseBroken.tiff \
  common_CloseBrokenH.tiff \
  common_CloseH.tiff \
- common_ClosedHandCursor.tiff \
  common_ColorSwatch.tiff \
  common_ComboBoxEllipsis.tiff \
  common_DecimalTabStop.tiff \
@@ -90,7 +89,6 @@
  common_MusicFolder.tiff \
  common_Mount.tiff \
  common_Nibble.tiff \
- common_OpenHandCursor.tiff \
  common_Printer.tiff \
  common_RadioOff.tiff \
  common_RadioOn.tiff \
@@ -172,6 +170,8 @@
  NSPreferencesGeneral.tiff \
  NSAdvanced.tiff \
  NSUserAccounts.tiff \
+ common_contextualMenuCursor.tiff \
+ common_disappearingItemCursor.tiff \
 
 
 include $(GNUSTEP_MAKEFILES)/resource-set.make
diff -Nru gnustep-gui-0.25.0.orig/Images/nsmapping.strings gnustep-gui-0.25.0/Images/nsmapping.strings
--- gnustep-gui-0.25.0.orig/Images/nsmapping.strings	2016-06-16 23:09:00.000000000 +0300
+++ gnustep-gui-0.25.0/Images/nsmapping.strings	2019-05-03 13:06:01.780677501 +0300
@@ -36,8 +36,11 @@
 common_ComboBoxEllipsis = GSComboArrow;
 NSComboArrow = GSComboArrow;
 
-common_ClosedHandCursor = GSClosedHandCursor;
-common_OpenHandCursor = GSOpenHandCursor;
+/* Mouse cursor images */
+
+common_contextualMenuCursor = GSContextualMenuCursor;
+common_disappearingItemCursor = GSDisappearingItemCursor;
+common_greenArrowCursor = GSGreenArrowCursor;
 common_copyCursor = GSDragCopyCursor;
 common_linkCursor = GSDragLinkCursor;
 common_noCursor = GSOperationNotAllowedCursor;
