diff -Nru gnustep-gui-0.25.0.orig/Source/GNUmakefile gnustep-gui-0.25.0/Source/GNUmakefile
--- gnustep-gui-0.25.0.orig/Source/GNUmakefile	2016-06-16 23:09:00.000000000 +0300
+++ gnustep-gui-0.25.0/Source/GNUmakefile	2019-05-03 17:34:18.250890506 +0300
@@ -23,7 +23,7 @@
 #  Boston, MA 02110-1301, USA.
 
 PACKAGE_NAME = gnustep-gui
-GNUSTEP_LOCAL_ADDITIONAL_MAKEFILES=../gui.make
+GNUSTEP_LOCAL_ADDITIONAL_MAKEFILES=../../gnustep-base-1.26.0/base.make ../gui.make
 include $(GNUSTEP_MAKEFILES)/common.make
 
 include ../Version
