Description: This patch includes the debian paths
Author: Rodolfo García Peñas (kix) <kix@kix.es>
Last-Update: 2011-08-16
Index: WindowMaker-0.95.6/doc/wmaker.1x
===================================================================
--- WindowMaker-0.95.6.orig/doc/wmaker.1x
+++ WindowMaker-0.95.6/doc/wmaker.1x
@@ -9,6 +9,9 @@ wmaker \- X11 window manager with a NEXT
 Window Maker is a X11 window manager with a NEXTSTEP look. It tries to
 emulate NeXT's look as much as possible, but it deviates from it as
 necessary.
+.PP
+\fIPlease note this manpage has been modified to reflect changes in the
+Debian installation of Window Maker.\fP
 .SH "OPTIONS"
 .TP
 .B \-display host:display.screen
@@ -78,7 +81,7 @@ Attribute Editor (right drag the applica
 Attributes) instead of modifying this file directly. There are just a
 few options not available using the Attributes Editor.
 .TP
-.B /usr/share/WindowMaker/Defaults/
+.B /etc/GNUstep/Defaults/
 All the above-mentioned files are READ from here if not found except
 for WMState, which is COPIED from here. No matter where they are read
 from, if it's necessary to write configuration changes back into this
@@ -106,6 +109,9 @@ and here, in that order. Unless, the ind
 .B ~/GNUstep/Library/WindowMaker/Pixmaps/
 Window Maker looks for \fBpixmaps\fP here
 .TP
+.B ~/GNUstep/Library/Icons/
+Window Maker looks for \fBicons\fP here
+.TP
 .B ~/GNUstep/Library/WindowMaker/Backgrounds/
 Window Maker looks for backgrounds here
 .TP
