diff -Nru gnustep-back-0.24.1.orig/Source/x11/XGServerWindow.m gnustep-back-0.24.1/Source/x11/XGServerWindow.m
--- gnustep-back-0.24.1.orig/Source/x11/XGServerWindow.m	2015-09-08 21:57:10.314108549 +0300
+++ gnustep-back-0.24.1/Source/x11/XGServerWindow.m	2015-09-08 22:06:14.454097446 +0300
@@ -3412,7 +3412,10 @@
               data[1] = generic.wintypes.win_menu_atom;
               len = 2;
 #else
-              data[0] = generic.wintypes.win_menu_atom;
+              if (level == NSFloatingWindowLevel)
+                data[0] = generic.wintypes.win_utility_atom;
+              else
+                data[0] = generic.wintypes.win_menu_atom;
               len = 1;
 #endif
               skipTaskbar = YES;
