diff --git a/CoreFoundation/Base.subproj/CFRuntime.c b/CoreFoundation/Base.subproj/CFRuntime.c
index 19167194..98aa3128 100644
--- a/CoreFoundation/Base.subproj/CFRuntime.c
+++ b/CoreFoundation/Base.subproj/CFRuntime.c
@@ -236,6 +236,7 @@ CFRuntimeClass const * __CFRuntimeClassTable[__CFRuntimeClassTableSize * 2] __at
 
 
 
+    [_kCFRuntimeIDCFNotificationCenter] = &__CFNotificationCenterClass,
     [_kCFRuntimeIDCFRunLoopMode] = &__CFRunLoopModeClass,
     [_kCFRuntimeIDCFRunLoop] = &__CFRunLoopClass,
     [_kCFRuntimeIDCFRunLoopSource] = &__CFRunLoopSourceClass,
diff --git a/CoreFoundation/Base.subproj/CFRuntime_Internal.h b/CoreFoundation/Base.subproj/CFRuntime_Internal.h
index 9a0345ff..b35fa0ae 100644
--- a/CoreFoundation/Base.subproj/CFRuntime_Internal.h
+++ b/CoreFoundation/Base.subproj/CFRuntime_Internal.h
@@ -109,6 +109,7 @@ CF_PRIVATE const CFRuntimeClass __CFPlugInInstanceClass;
 
 
 CF_PRIVATE const CFRuntimeClass __CFPasteboardClass;
+CF_PRIVATE const CFRuntimeClass __CFNotificationCenterClass;
 CF_PRIVATE const CFRuntimeClass __CFUserNotificationClass;
 CF_PRIVATE const CFRuntimeClass __CFUUIDClass;
 CF_PRIVATE const CFRuntimeClass __CFLocaleClass;
diff --git a/CoreFoundation/CMakeLists.txt b/CoreFoundation/CMakeLists.txt
index 63dafee9..a063afe0 100644
--- a/CoreFoundation/CMakeLists.txt
+++ b/CoreFoundation/CMakeLists.txt
@@ -248,6 +248,8 @@ add_framework(CoreFoundation
                 URL.subproj/CFURLAccess.h
                 URL.subproj/CFURLComponents.h
               SOURCES
+                # AppServices
+                AppServices.subproj/CFNotificationCenter.c
                 # Base
                 Base.subproj/CFBase.c
                 Base.subproj/CFFileUtilities.c
