--- ./WindowMaker-0.95.7/src/screen.h	2015-08-11 21:41:14.000000000 +0300
+++ ./WindowMaker/src/screen.h	2016-11-25 00:09:13.798571299 +0200
@@ -250,6 +250,9 @@
     WMScreen *wmscreen;		       /* for widget library */
 
     struct RImage *icon_tile;
+#ifdef NEXTSPACE
+    struct RImage *miniwindow_tile;
+#endif
     struct RImage *clip_tile;          /* tile with arrows to change workspace */
     struct RImage *drawer_tile;        /* tile for a drawer (tile + arrow) */
     Pixmap icon_tile_pixmap;		/* For app supplied icons */
