--- ./WindowMaker-0.95.7/src/event.c	2015-08-11 21:41:14.000000000 +0300
+++ ./WindowMaker/src/event.c	2018-07-24 00:42:48.434242727 +0300
@@ -80,6 +80,13 @@
 #include "switchmenu.h"
 #include "wsmap.h"
 
+#ifdef NEXTSPACE
+#include <Workspace+WindowMaker.h>
+extern void WWMIconYardShowIcons(WScreen *screen);
+extern void WWMIconYardHideIcons(WScreen *screen);
+extern void WWMDockShowIcons(WDock *dock);
+extern void WWMDockHideIcons(WDock *dock);
+#endif
 
 #define MOD_MASK wPreferences.modifier_mask
 
@@ -90,6 +97,10 @@
 static void handleMapNotify(XEvent *event);
 static void handleUnmapNotify(XEvent *event);
 static void handleButtonPress(XEvent *event);
+#ifdef NEXTSPACE
+static void handleButtonRelease(XEvent * event);
+static void handleKeyRelease(XEvent * event);
+#endif
 static void handleExpose(XEvent *event);
 static void handleDestroyNotify(XEvent *event);
 static void handleConfigureRequest(XEvent *event);
@@ -209,6 +220,11 @@
 		handleKeyPress(event);
 		break;
 
+#ifdef NEXTSPACE
+	case KeyRelease:
+		handleKeyRelease(event);
+		break;
+#endif
 	case MotionNotify:
 		handleMotionNotify(event);
 		break;
@@ -237,6 +253,11 @@
 		handleExpose(event);
 		break;
 
+#ifdef NEXTSPACE
+	case ButtonRelease:
+		handleButtonRelease(event);
+          break;
+#endif
 	case PropertyNotify:
 		handlePropertyNotify(event);
 		break;
@@ -580,9 +601,15 @@
 		 * XRRUpdateConfiguration when screen configuration change notify
 		 * events are generated */
 		XRRUpdateConfiguration(event);
+#ifdef NEXTSPACE                
+		for (int i = 0; i < w_global.screen_count; i++) {
+			XWUpdateScreenInfo(wScreenWithNumber(i));
+		}
+#else
 		WCHANGE_STATE(WSTATE_RESTARTING);
 		Shutdown(WSRestartPreparationMode);
 		Restart(NULL,True);
+#endif
 	}
 #endif
 }
@@ -670,6 +697,9 @@
 
 	wwin = wWindowFor(window);
 	if (wwin) {
+#ifdef NEXTSPACE
+		dispatch_sync(workspace_q, ^{ XWApplicationDidCloseWindow(wwin); });
+#endif
 		wUnmanageWindow(wwin, False, True);
 	}
 
@@ -803,17 +833,48 @@
 
 	scr = wScreenForRootWindow(event->xbutton.root);
 
+#ifdef NEXTSPACE
+  // reset current focused window button beacuse ButtonPress may change focus
+	WWindow *wwin = scr->focused_window;
+	if (wwin && wwin->client_win != scr->no_focus_win &&
+			wwin->frame->left_button &&
+			event->xbutton.window != wwin->frame->left_button->window &&
+			wwin->frame->right_button &&
+			event->xbutton.window != wwin->frame->right_button->window) {
+		scr->flags.modifier_pressed = 0;
+		wWindowUpdateButtonImages(wwin);
+	}
+#endif
+
 #ifdef BALLOON_TEXT
 	wBalloonHide(scr);
 #endif
 
 	if (!wPreferences.disable_root_mouse && event->xbutton.window == scr->root_win) {
 		if (event->xbutton.button == Button1 && wPreferences.mouse_button1 != WA_NONE) {
+#ifdef NEXTSPACE
+			if (scr->focused_window && scr->focused_window->flags.is_gnustep) {
+				XSendEvent(dpy, scr->focused_window->client_win, False, ButtonPressMask, event);
+			}
+			else {
+				XSendEvent(dpy, scr->dock->icon_array[0]->icon->icon_win, False, ButtonPressMask, event);
+			}
+#else
 			executeButtonAction(scr, event, wPreferences.mouse_button1);
+#endif
 		} else if (event->xbutton.button == Button2 && wPreferences.mouse_button2 != WA_NONE) {
 			executeButtonAction(scr, event, wPreferences.mouse_button2);
 		} else if (event->xbutton.button == Button3 && wPreferences.mouse_button3 != WA_NONE) {
+#ifdef NEXTSPACE
+			if (scr->focused_window && scr->focused_window->flags.is_gnustep) {
+				XSendEvent(dpy, scr->focused_window->client_win, False, ButtonPressMask, event);
+			}
+			else {
+				XSendEvent(dpy, scr->dock->icon_array[0]->icon->icon_win, False, ButtonPressMask, event);
+			}
+#else
 			executeButtonAction(scr, event, wPreferences.mouse_button3);
+#endif
 		} else if (event->xbutton.button == Button8 && wPreferences.mouse_button8 != WA_NONE) {
 			executeButtonAction(scr, event, wPreferences.mouse_button8);
 		}else if (event->xbutton.button == Button9 && wPreferences.mouse_button9 != WA_NONE) {
@@ -839,15 +900,12 @@
 	if (desc->parent_type == WCLASS_WINDOW) {
 		XSync(dpy, 0);
 
-		if (event->xbutton.state & ( MOD_MASK | ControlMask )) {
+		if (event->xbutton.state & ( MOD_MASK | ControlMask)) {
+			XAllowEvents(dpy, AsyncPointer, CurrentTime);
+		} else if (wPreferences.ignore_focus_click) {
 			XAllowEvents(dpy, AsyncPointer, CurrentTime);
 		} else {
-			/*      if (wPreferences.focus_mode == WKF_CLICK) { */
-			if (wPreferences.ignore_focus_click) {
-				XAllowEvents(dpy, AsyncPointer, CurrentTime);
-			}
 			XAllowEvents(dpy, ReplayPointer, CurrentTime);
-			/*      } */
 		}
 		XSync(dpy, 0);
 	} else if (desc->parent_type == WCLASS_APPICON
@@ -873,6 +931,23 @@
 	}
 }
 
+#ifdef NEXTSPACE
+static void handleButtonRelease(XEvent * event)
+{
+	WScreen *scr = wScreenForRootWindow(event->xbutton.root);
+
+	if (!wPreferences.disable_root_mouse && event->xbutton.window == scr->root_win
+			&& event->xbutton.button == Button3) {
+		if (scr->focused_window && scr->focused_window->flags.is_gnustep) {
+			XSendEvent(dpy, scr->focused_window->client_win, True, ButtonReleaseMask, event);
+		}
+		else {
+			XSendEvent(dpy, scr->dock->icon_array[0]->icon->icon_win, False, ButtonReleaseMask, event);
+		}
+	}
+}
+#endif
+
 static void handleMapNotify(XEvent * event)
 {
 	WWindow *wwin;
@@ -1389,20 +1464,42 @@
 	/* ignore CapsLock */
 	modifiers = event->xkey.state & w_global.shortcut.modifiers_mask;
 
+#ifdef NEXTSPACE
+  if (wwin && wwin->client_win) {
+    fprintf(stderr, "[WindowMaker] handleKeyPress: %i state: %i mask: %i"
+            " modifiers: %i window:%lu\n",
+            event->xkey.keycode, event->xkey.state, MOD_MASK,
+            modifiers, wwin->client_win);
+  }
+  
+	if (((event->xkey.keycode == XKeysymToKeycode(dpy, XK_Super_L)) ||
+       (event->xkey.keycode == XKeysymToKeycode(dpy, XK_Super_R))) &&
+			modifiers == 0) {
+    if (wwin && wwin->client_win != scr->no_focus_win &&
+        event->xkey.window != event->xkey.root) {
+      scr->flags.modifier_pressed = 1;
+      wWindowUpdateButtonImages(wwin);
+    }
+	}
+	else if (event->xkey.window != event->xkey.root &&
+           event->xkey.window != scr->no_focus_win) {
+		scr->flags.modifier_pressed = 0;
+		wWindowUpdateButtonImages(wwin);
+	}
+#endif
+
 	for (i = 0; i < WKBD_LAST; i++) {
 		if (wKeyBindings[i].keycode == 0)
 			continue;
 
-		if (wKeyBindings[i].keycode == event->xkey.keycode && (	/*wKeyBindings[i].modifier==0
-									   || */ wKeyBindings[i].modifier ==
-									      modifiers)) {
+		if (wKeyBindings[i].keycode == event->xkey.keycode &&
+				(wKeyBindings[i].modifier == modifiers)) {
 			command = i;
 			break;
 		}
 	}
 
 	if (command < 0) {
-
 		if (!wRootMenuPerformShortcut(event)) {
 			static int dontLoop = 0;
 
@@ -1420,6 +1517,15 @@
 			}
 			dontLoop--;
 		}
+
+    // Shortuct which does not overlap with WindowMaker was pressed -
+    // send it to GNUstep application. For example, Alternate-x
+    // pressed over Terminal window which runs Emacs should result in
+    // appearing 'M-x' prompt in Emacs.
+    if (wwin && wwin->flags.is_gnustep) {
+      XSendEvent(dpy, wwin->client_win, True, KeyPress, event);
+    }
+    
 		return;
 	}
 #define ISMAPPED(w) ((w) && !(w)->flags.miniaturized && ((w)->flags.mapped || (w)->flags.shaded))
@@ -1427,6 +1533,26 @@
 
 	switch (command) {
 
+  case WKBD_DOCKHIDESHOW:
+    if (strcmp(scr->focused_window->wm_instance, "Workspace") != 0) {
+      if (scr->dock->mapped) {
+        WWMDockHideIcons(scr->dock);
+      }
+      else {
+        WWMDockShowIcons(scr->dock);
+      }
+    }
+    break;
+  case WKBD_YARDHIDESHOW:
+    if (strcmp(scr->focused_window->wm_instance, "Workspace") != 0) {
+      if (scr->flags.icon_yard_mapped) {
+        WWMIconYardHideIcons(scr);
+      }
+      else {
+        WWMIconYardShowIcons(scr);
+      }
+    }
+    break;
 	case WKBD_ROOTMENU:
 		/*OpenRootMenu(scr, event->xkey.x_root, event->xkey.y_root, True); */
 		if (!CheckFullScreenWindowFocused(scr)) {
@@ -1452,12 +1578,13 @@
 		wHideAll(scr);
 		break;
 	case WKBD_MINIATURIZE:
-		if (ISMAPPED(wwin) && ISFOCUSED(wwin)
-		    && !WFLAGP(wwin, no_miniaturizable)) {
+		if (ISMAPPED(wwin) && ISFOCUSED(wwin) && !WFLAGP(wwin, no_miniaturizable)) {
 			CloseWindowMenu(scr);
-
-			if (wwin->protocols.MINIATURIZE_WINDOW)
-				wClientSendProtocol(wwin, w_global.atom.gnustep.wm_miniaturize_window, event->xbutton.time);
+			if (wwin->protocols.MINIATURIZE_WINDOW) {
+        fprintf(stderr, "[WM] send WM_MINIATURIZE_WINDOW protocol message to client.\n");
+        wClientSendProtocol(wwin, w_global.atom.gnustep.wm_miniaturize_window,
+                            event->xbutton.time);
+      }
 			else {
 				wIconifyWindow(wwin);
 			}
@@ -1469,8 +1596,20 @@
 			CloseWindowMenu(scr);
 
 			if (wapp && !WFLAGP(wapp->main_window_desc, no_appicon)) {
-				wHideApplication(wapp);
-			}
+        if (wwin->protocols.HIDE_APP) {
+          WIcon *icon = wapp->app_icon->icon;
+          fprintf(stderr, "[WM] send WM_HIDE_APP protocol message to client.\n");
+          animateResize(wwin->screen_ptr, wwin->frame_x, wwin->frame_y,
+                        wwin->frame->core->width, wwin->frame->core->height,
+                        wapp->app_icon->x_pos, wapp->app_icon->y_pos,
+                        icon->core->width, icon->core->height);
+          wClientSendProtocol(wwin, w_global.atom.gnustep.wm_hide_app,
+                              event->xbutton.time);
+        }
+        else {
+          wHideApplication(wapp);
+        }
+      }
 		}
 		break;
 	case WKBD_HIDE_OTHERS:
@@ -1866,11 +2005,59 @@
 #endif	/* KEEP_XKB_LOCK_STATUS */
 	}
 }
-
+#ifdef NEXTSPACE
+static void handleKeyRelease(XEvent * event)
+{
+	WScreen *scr = wScreenForRootWindow(event->xkey.root);
+	WWindow *wwin = scr->focused_window;
+  
+  if (event->xkey.window == event->xkey.root ||
+      event->xkey.window == scr->no_focus_win) {
+    return;
+  }
+  /* fprintf(stderr, "[WindowMaker] handleKeyRelease: %i state: %i mask: %i\n", */
+  /*         event->xkey.keycode, event->xkey.state, MOD_MASK); */
+	if ( (event->xkey.keycode == XKeysymToKeycode(dpy, XK_Super_L)) ||
+       (event->xkey.keycode == XKeysymToKeycode(dpy, XK_Super_R)) ) {
+    if (wwin) {
+      scr->flags.modifier_pressed = 0;
+      wWindowUpdateButtonImages(wwin);
+      if (wwin->flags.is_gnustep) {
+        XSendEvent(dpy, scr->focused_window->client_win, True, KeyRelease, event);
+      }
+    }
+  }
+}
+#endif
 static void handleMotionNotify(XEvent * event)
 {
 	WScreen *scr = wScreenForRootWindow(event->xmotion.root);
 
+#ifdef NEXTSPACE
+	WWindow *wwin = wWindowFor(event->xmotion.window);
+
+	if (event->xmotion.state == 0 || wwin == NULL) {
+		return;
+	}
+
+	if (event->xmotion.state & Button1Mask &&
+			XGrabPointer(dpy, event->xmotion.window, False,
+									ButtonMotionMask | ButtonReleaseMask | ButtonPressMask,
+									GrabModeAsync, GrabModeAsync, None, None, CurrentTime) == GrabSuccess) {
+		/* wMouseMoveWindow checks for button on ButtonRelease event inside it's loop */
+		event->xbutton.button = Button1;
+		if (event->xmotion.window == wwin->frame->titlebar->window ||
+				event->xmotion.state & MOD_MASK) {
+			/* move the window */
+			wMouseMoveWindow(wwin, event);
+		}
+		else if (IS_RESIZABLE(wwin) && event->xmotion.window == wwin->frame->resizebar->window) {
+			wMouseResizeWindow(wwin, event);
+		}
+		XUngrabPointer(dpy, CurrentTime);
+	}
+#endif
+
 	if (wPreferences.scrollable_menus) {
 		WMPoint p = wmkpoint(event->xmotion.x_root, event->xmotion.y_root);
 		WMRect rect = wGetRectForHead(scr, wGetHeadForPoint(scr, p));
