--- ./WindowMaker-0.95.7/src/def_pixmaps.h	2015-08-11 21:41:14.000000000 +0300
+++ ./WindowMaker/src/def_pixmaps.h	2018-06-25 18:10:38.256253999 +0300
@@ -46,7 +46,6 @@
     ". .XXXX. .",
     " .XXXXXX. "};
 
-
 static char *PRED_BROKEN_CLOSE_XPM[] = {
     "10 10 3 1",
     " 	c #000000",
@@ -63,26 +62,42 @@
     ". .XXXX. .",
     " .XXXXXX. "};
 
+/* static char *PRED_KILL_XPM[] = { */
+/*     "10 10 3 1", */
+/*     " 	c #000000", */
+/*     ".	c #616561", */
+/*     "X	c None", */
+/*     " .XXXXXX. ", */
+/*     ". XXXXXX .", */
+/*     "XXXXXXXXXX", */
+/*     "XXX .. XXX", */
+/*     "XXX.  .XXX", */
+/*     "XXX.  .XXX", */
+/*     "XXX .. XXX", */
+/*     "XXXXXXXXXX", */
+/*     ". XXXXXX .", */
+/*     " .XXXXXX. "}; */
 
 static char *PRED_KILL_XPM[] = {
-    "10 10 3 1",
-    " 	c #000000",
-    ".	c #616561",
-    "X	c None",
-    " .XXXXXX. ",
-    ". XXXXXX .",
-    "XXXXXXXXXX",
-    "XXX .. XXX",
-    "XXX.  .XXX",
-    "XXX.  .XXX",
-    "XXX .. XXX",
-    "XXXXXXXXXX",
-    ". XXXXXX .",
-    " .XXXXXX. "};
+"10 10 4 1",
+" 	c None",
+".	c #555555",
+"+	c #000000",
+"@	c #777777",
+".+  ..  +.",
+"++..++..++",
+" .++++++. ",
+" .+@++@+. ",
+" .+ ++ +. ",
+" .++++++. ",
+"  .+  +.  ",
+" .+.++.+. ",
+"++.    .++",
+".+      +."};
 
 
 static char *PRED_ICONIFY_XPM[] = {
-    "    10    10        2            1",
+    "10 10 2 1",
     ". c #000000",
     "# c None",
     "..........",
@@ -97,6 +112,39 @@
     ".........."
 };
 
+static char *PRED_MAXIMIZE_XPM[] = {
+"10 10 3 1",
+" 	c None",
+".	c #000000",
+"+	c #555555",
+"...+  +...",
+"..+    +..",
+".+.+  +.+.",
+"+ +.  .+ +",
+"          ",
+"          ",
+"+ +.  .+ +",
+".+.+  +.+.",
+"..+    +..",
+"...+  +..."};
+
+static char *PRED_RESTORE_xpm[] = {
+"10 10 3 1",
+" 	c None",
+".	c #000000",
+"+	c #555555",
+".+ +  + +.",
+"+.+.  .+.+",
+" +..  ..+ ",
+"+...  ...+",
+"          ",
+"          ",
+"+...  ...+",
+" +..  ..+ ",
+"+.+.  .+.+",
+".+ +  + +."};
+
+
 static char *NEXT_CLOSE_XPM[] = {
 "10 10 69 1",
 " 	c None",
@@ -179,7 +227,6 @@
 "%wxx'-p8tn",
 "yzAB7C8enD"};
 
-
 static char *NEXT_BROKEN_CLOSE_XPM[] = {
 "10 10 8 1",
 " 	c None",
