--- ./WindowMaker-0.95.7/src/screen.c	2015-08-11 21:41:14.000000000 +0300
+++ ./WindowMaker/src/screen.c	2018-01-22 00:48:54.198616011 +0200
@@ -898,6 +898,7 @@
 	if (!scr->session_state)
 		scr->session_state = WMCreatePLDictionary(NULL, NULL);
 
+#ifndef NEXTSPACE
 	if (!wPreferences.flags.nodock) {
 		state = WMGetFromPLDictionary(scr->session_state, dDock);
 		scr->dock = wDockRestoreState(scr, state, WM_DOCK);
@@ -919,6 +920,7 @@
 		}
 		wDrawersRestoreState(scr);
 	}
+#endif
 
 	wWorkspaceRestoreState(scr);
 	wScreenUpdateUsableArea(scr);
