--- ./WindowMaker-0.95.7/src/wconfig.h.in	2015-08-11 21:41:14.000000000 +0300
+++ ./WindowMaker/src/wconfig.h.in	2018-08-10 17:08:28.874128500 +0300
@@ -122,7 +122,7 @@
 /* name of the script to execute at startup */
 #define DEF_INIT_SCRIPT "autostart"
 #define DEF_EXIT_SCRIPT "exitscript"
-#define DEFAULTS_DIR    "Defaults"
+#define DEFAULTS_DIR    "Preferences/.WindowMaker"
 
 #ifdef USE_TIFF
 #define DEF_BUTTON_IMAGES PKGDATADIR"/buttons.tiff"
@@ -146,7 +146,7 @@
 
 /* icon path */
 #define DEF_ICON_PATHS \
-    "(\"~/pixmaps\",\"~/GNUstep/Library/Icons\",\"/usr/include/X11/pixmaps/\",\""PIXMAPDIR"\")"
+    "(\"~/pixmaps\",\"~/Library/Images\",\"/usr/include/X11/pixmaps/\",\""PIXMAPDIR"\")"
 
 /* window title to use for untitled windows */
 #define DEF_WINDOW_TITLE "Untitled"
@@ -155,13 +155,13 @@
 #define DEF_FRAME_COLOR   "white"
 
 /* default fonts */
-#define DEF_TITLE_FONT          "\"sans-serif:bold:pixelsize=12\""
-#define DEF_MENU_TITLE_FONT     "\"sans-serif:bold:pixelsize=12\""
-#define DEF_MENU_ENTRY_FONT     "\"sans-serif:pixelsize=12\""
-#define DEF_ICON_TITLE_FONT     "\"sans-serif:pixelsize=9\""
-#define DEF_CLIP_TITLE_FONT     "\"sans-serif:bold:pixelsize=10\""
-#define DEF_INFO_TEXT_FONT      "\"sans-serif:pixelsize=12\""
-#define DEF_WORKSPACE_NAME_FONT "\"sans-serif:pixelsize=24\""
+#define DEF_TITLE_FONT          "\"Helvetica:bold:pixelsize=12\""
+#define DEF_MENU_TITLE_FONT     "\"Helvetica:bold:pixelsize=12\""
+#define DEF_MENU_ENTRY_FONT     "\"Helvetica:pixelsize=12\""
+#define DEF_ICON_TITLE_FONT     "\"Helvetica:pixelsize=9\""
+#define DEF_CLIP_TITLE_FONT     "\"Helvetica:bold:pixelsize=10\""
+#define DEF_INFO_TEXT_FONT      "\"Helvetica:pixelsize=12\""
+#define DEF_WORKSPACE_NAME_FONT "\"Helvetica:pixelsize=24\""
 
 /* line width of the move/resize frame */
 #define DEF_FRAME_THICKNESS              1
@@ -282,7 +282,7 @@
 #define PLACETEST_HSTEP	        8
 #define PLACETEST_VSTEP	        8
 
-#define DOCK_EXTRA_SPACE	0
+#define DOCK_EXTRA_SPACE	3
 
 /* Vicinity in which an icon can be attached to the clip */
 #define CLIP_ATTACH_VICINITY	1
@@ -290,14 +290,14 @@
 
 /* The amount of space (in multiples of the icon size)
  * a docked icon must be dragged out to detach it */
-#define DOCK_DETTACH_THRESHOLD	3
+#define DOCK_DETTACH_THRESHOLD	2
 
 /* Max. number of icons the dock and clip can have */
 #define DOCK_MAX_ICONS		32
 
 /* blink interval when invoking a menu item */
 #define MENU_BLINK_DELAY	60000
-#define MENU_BLINK_COUNT	2
+#define MENU_BLINK_COUNT	0
 #define CURSOR_BLINK_RATE	300
 
 /* how many pixels to move before dragging windows and other objects */
