--- ./WindowMaker-0.95.7/src/WindowMaker.h	2015-08-11 21:41:14.000000000 +0300
+++ ./WindowMaker/src/WindowMaker.h	2018-06-25 16:59:51.695034500 +0300
@@ -56,7 +56,7 @@
 	WMNormalLevel = 0,
 	WMFloatingLevel = 3,
 	WMDockLevel = 5,
-	WMSubmenuLevel = 15,
+	WMSubmenuLevel = 10,
 	WMMainMenuLevel = 20,
 	WMStatusLevel = 21,
 	WMModalLevel = 100,
@@ -92,14 +92,16 @@
 #define WBUT_BROKENCLOSE        1
 #define WBUT_ICONIFY            2
 #define WBUT_KILL		3
+#define WBUT_MAXIMIZE		4
+#define WBUT_RESTORE		5
 #ifdef XKB_BUTTON_HINT
-#define WBUT_XKBGROUP1		4
-#define WBUT_XKBGROUP2		5
-#define WBUT_XKBGROUP3		6
-#define WBUT_XKBGROUP4		7
-#define PRED_BPIXMAPS		8 /* reserved for 4 groups */
+#define WBUT_XKBGROUP1		6
+#define WBUT_XKBGROUP2		7
+#define WBUT_XKBGROUP3		8
+#define WBUT_XKBGROUP4		9
+#define PRED_BPIXMAPS		10 /* reserved for 4 groups */
 #else
-#define PRED_BPIXMAPS		4 /* count of WBUT icons */
+#define PRED_BPIXMAPS		6 /* count of WBUT icons */
 #endif /* XKB_BUTTON_HINT */
 
 /* Mouse cursors */
@@ -113,6 +115,10 @@
 	WCUR_BOTTOMRIGHTRESIZE,
 	WCUR_VERTICALRESIZE,
 	WCUR_HORIZONRESIZE,
+	WCUR_UPRESIZE,
+	WCUR_DOWNRESIZE,
+	WCUR_LEFTRESIZE,
+	WCUR_RIGHTRESIZE,
 	WCUR_WAIT,
 	WCUR_ARROW,
 	WCUR_QUESTION,
