--- ./WindowMaker-0.95.7/src/dialog.c	2015-08-11 21:41:14.000000000 +0300
+++ ./WindowMaker/src/dialog.c	2016-11-25 00:09:13.797571299 +0200
@@ -177,7 +177,11 @@
 	char *filename = NULL;
 
 	filename = wstrdup(wusergnusteppath());
+#ifdef NEXTSPACE
+	filename = wstrappend(filename, "/WindowMaker/History");
+#else
 	filename = wstrappend(filename, "/.AppInfo/WindowMaker/History");
+#endif
 	if (name && strlen(name)) {
 		filename = wstrappend(filename, ".");
 		filename = wstrappend(filename, name);
