--- ./wmaker-crm_next/src/main.c	2015-07-29 16:02:49.000000000 +0300
+++ ./WM/src/main.c	2015-08-25 02:17:07.278585171 +0300
@@ -105,7 +105,7 @@
 static int *wVisualID = NULL;
 static int wVisualID_len = 0;
 
-static int real_main(int argc, char **argv);
+int real_main(int argc, char **argv);
 
 int getWVisualID(int screen)
 {
@@ -492,7 +492,11 @@
 			   " Changes to the defaults database will require"
 			   " a restart to take effect. Check your kernel!"));
 	} else {
+#ifdef NEXTSPACE
+		watchPath = wdefaultspathfordomain("");
+#else
 		watchPath = wstrconcat(wusergnusteppath(), "/Defaults");
+#endif
 		/* Add the watch; really we are only looking for modify events
 		 * but we might want more in the future so check all events for now.
 		 * The individual events are checked for in event.c.
@@ -514,7 +518,11 @@
 {
 	char *file, *paths;
 
+#ifdef NEXTSPACE
+	paths = wstrconcat(wusergnusteppath(), "/WindowMaker");
+#else
 	paths = wstrconcat(wusergnusteppath(), "/Library/WindowMaker");
+#endif
 	paths = wstrappend(paths, ":" DEF_CONFIG_PATHS);
 
 	file = wfindfile(paths, DEF_INIT_SCRIPT);
@@ -546,6 +554,7 @@
 	}
 }
 
+#ifndef NEXTSPACE
 int main(int argc, char **argv)
 {
 	int i_am_the_monitor, i, len;
@@ -604,8 +613,9 @@
 	else
 		return real_main(argc, argv);
 }
+#endif // NEXTSPACE
 
-static int real_main(int argc, char **argv)
+int real_main(int argc, char **argv)
 {
 	int i;
 	char *pos;
@@ -808,6 +818,10 @@
 #ifdef HAVE_INOTIFY
 	inotifyWatchConfig();
 #endif
+#ifdef NEXTSPACE
+	return 0;
+#else
 	EventLoop();
 	return -1;
+#endif // NEXTSPACE
 }
