--- ./WindowMaker-0.95.7/src/workspace.c	2015-08-11 21:41:14.000000000 +0300
+++ ./WindowMaker/src/workspace.c	2018-05-27 01:37:59.829006990 +0300
@@ -50,6 +50,9 @@
 #include "xinerama.h"
 #include "event.h"
 #include "wsmap.h"
+#ifdef NEXTSPACE
+#include <Workspace+WindowMaker.h>
+#endif // NEXTSPACE        
 
 #define MC_NEW          0
 #define MC_DESTROY_LAST 1
@@ -436,6 +439,9 @@
 
 	if (workspace != scr->current_workspace)
 		wWorkspaceForceChange(scr, workspace);
+#ifdef NEXTSPACE
+	XWMWorkspaceDidChange(scr, workspace);
+#endif // NEXTSPACE        
 }
 
 void wWorkspaceRelativeChange(WScreen * scr, int amount)
