--- ./wmaker-crm_next/src/event.c	2015-07-29 16:02:49.000000000 +0300
+++ ./WM/src/event.c	2015-08-22 22:54:58.000000000 +0300
@@ -80,6 +80,7 @@
 #include "switchmenu.h"
 #include "wsmap.h"
 
+#include <Workspace+WindowMaker.h>
 
 #define MOD_MASK wPreferences.modifier_mask
 
@@ -669,7 +670,15 @@
 	int widx;
 
 	wwin = wWindowFor(window);
+        
+       
 	if (wwin) {
+#ifdef NEXTSPACE
+		dispatch_sync(workspace_q,
+		    ^{
+			XWApplicationDidCloseWindow(wwin);
+		    });
+#endif
 		wUnmanageWindow(wwin, False, True);
 	}
 
