--- ./WindowMaker-0.95.7/src/dock.c	2015-08-11 21:41:14.000000000 +0300
+++ ./WindowMaker/src/dock.c	2018-01-22 18:12:44.020000000 +0200
@@ -854,7 +854,9 @@
 	btn->icon->core->descriptor.handle_leavenotify = clipLeaveNotify;
 	btn->icon->core->descriptor.parent_type = WCLASS_DOCK_ICON;
 	btn->icon->core->descriptor.parent = btn;
+#ifndef NEXTSPACE
 	XMapWindow(dpy, btn->icon->core->window);
+#endif
 	btn->x_pos = x_pos;
 	btn->y_pos = 0;
 	btn->docked = 1;
@@ -1959,9 +1961,10 @@
 				ChangeStackingLevel(aicon->icon->core, WMDockLevel);
 
 			wCoreConfigure(aicon->icon->core, aicon->x_pos, aicon->y_pos, 0, 0);
+#ifndef NEXTSPACE
 			if (!dock->collapsed)
 				XMapWindow(dpy, aicon->icon->core->window);
-
+#endif
 			wRaiseFrame(aicon->icon->core);
 
 			dock->icon_count++;
