--- ./WindowMaker-0.95.7/src/WindowMaker.h	2015-08-11 21:41:14.000000000 +0300
+++ ./WindowMaker/src/WindowMaker.h	2018-02-15 16:28:22.843424004 +0200
@@ -56,7 +56,7 @@
 	WMNormalLevel = 0,
 	WMFloatingLevel = 3,
 	WMDockLevel = 5,
-	WMSubmenuLevel = 15,
+	WMSubmenuLevel = 10,
 	WMMainMenuLevel = 20,
 	WMStatusLevel = 21,
 	WMModalLevel = 100,
@@ -113,6 +113,10 @@
 	WCUR_BOTTOMRIGHTRESIZE,
 	WCUR_VERTICALRESIZE,
 	WCUR_HORIZONRESIZE,
+	WCUR_UPRESIZE,
+	WCUR_DOWNRESIZE,
+	WCUR_LEFTRESIZE,
+	WCUR_RIGHTRESIZE,
 	WCUR_WAIT,
 	WCUR_ARROW,
 	WCUR_QUESTION,
