--- ./WindowMaker-0.95.7/src/defaults.c	2015-08-11 21:41:14.000000000 +0300
+++ ./WindowMaker/src/defaults.c	2018-02-15 16:36:38.038898005 +0200
@@ -118,6 +118,9 @@
 static WDECallbackUpdate setStickyIcons;
 static WDECallbackUpdate setWidgetColor;
 static WDECallbackUpdate setIconTile;
+#ifdef NEXTSPACE
+static WDECallbackUpdate setMiniwindowTile;
+#endif
 static WDECallbackUpdate setWinTitleFont;
 static WDECallbackUpdate setMenuTitleFont;
 static WDECallbackUpdate setMenuTextFont;
@@ -553,6 +556,10 @@
 	    NULL, getBool, NULL, NULL, NULL},
 	{"IconBack", "(solid, gray)", NULL,
 	    NULL, getTexture, setIconTile, NULL, NULL},
+#ifdef NEXTSPACE
+	{"MiniwindowBack", "(solid, gray)", NULL,
+	    NULL, getTexture, setMiniwindowTile, NULL, NULL},
+#endif
 	{"TitleJustify", "center", seJustifications,
 	    &wPreferences.title_justification, getEnum, setJustify, NULL, NULL},
 	{"WindowTitleFont", DEF_TITLE_FONT, NULL,
@@ -818,6 +825,14 @@
 	    NULL, getCursor, setCursor, NULL, NULL},
 	{"HorizontalResizeCursor", "(builtin, sb_h_double_arrow)", (void *)WCUR_HORIZONRESIZE,
 	    NULL, getCursor, setCursor, NULL, NULL},
+	{"UpResizeCursor", "(builtin, sb_up_arrow)", (void *)WCUR_UPRESIZE,
+	    NULL, getCursor, setCursor, NULL, NULL},
+	{"DownResizeCursor", "(builtin, sb_down_arrow)", (void *)WCUR_DOWNRESIZE,
+	    NULL, getCursor, setCursor, NULL, NULL},
+	{"LeftResizeCursor", "(builtin, sb_left_arrow)", (void *)WCUR_LEFTRESIZE,
+	    NULL, getCursor, setCursor, NULL, NULL},
+	{"RightResizeCursor", "(builtin, sb_right_arrow)", (void *)WCUR_RIGHTRESIZE,
+	    NULL, getCursor, setCursor, NULL, NULL},
 	{"WaitCursor", "(builtin, watch)", (void *)WCUR_WAIT,
 	    NULL, getCursor, setCursor, NULL, NULL},
 	{"QuestionCursor", "(builtin, question_arrow)", (void *)WCUR_QUESTION,
@@ -2379,6 +2394,9 @@
 	}
 }
 
+#ifdef NEXTSPACE
+#include <X11/Xcursor/Xcursor.h>
+#endif
 /*
  * (none)
  * (builtin, <cursor_name>)
@@ -2482,6 +2500,26 @@
 		wfree(bitmap_name);
 		wfree(mask_name);
 	}
+#ifdef NEXTSPACE
+	else if (strcasecmp(val, "library") == 0) {
+		if (nelem != 2) {
+			wwarning(_("bad number of arguments in cursor specification"));
+			return (status);
+		}
+		elem = WMGetFromPLArray(pl, 1);
+		if (!elem || !WMIsPLString(elem)) {
+			return (status);
+		}
+		val = WMGetFromPLString(elem);
+
+		*cursor = XcursorLibraryLoadCursor(dpy, val);
+		status = 1;
+
+		if (cursor == NULL) {
+			wwarning(_("unknown builtin cursor name \"%s\""), val);
+		}
+	}
+#endif
 	return (status);
 }
 
@@ -2681,6 +2719,56 @@
 	return (reset ? REFRESH_ICON_TILE : 0);
 }
 
+#ifdef NEXTSPACE
+static int setMiniwindowTile(WScreen * scr, WDefaultEntry * entry, void *tdata, void *foo)
+{
+  Pixmap	pixmap;
+  RImage	*img;
+  WTexture	**texture = tdata;
+  int		reset = 0;
+
+  /* Parameter not used, but tell the compiler that it is ok */
+  (void) foo;
+
+  img = wTextureRenderImage(*texture, wPreferences.icon_size,
+                            wPreferences.icon_size,
+                            ((*texture)->any.type & WREL_BORDER_MASK)
+                            ? WREL_ICON : WREL_FLAT);
+  if (!img)
+    {
+      wwarning(_("could not render texture for miniwindow background"));
+      if (!entry->addr)
+        wTextureDestroy(scr, *texture);
+      return 0;
+    }
+  RConvertImage(scr->rcontext, img, &pixmap);
+
+  if (scr->miniwindow_tile)
+    {
+      reset = 1;
+      RReleaseImage(scr->miniwindow_tile);
+    }
+
+  scr->miniwindow_tile = img;
+
+  /* put the icon in the noticeboard hint */
+  /* PropSetIconTileHint(scr, img); */
+
+  /* scr->icon_tile_pixmap = pixmap; */
+
+  /* icon back color for shadowing */
+  /*  if (scr->icon_back_texture)
+    wTextureDestroy(scr, (WTexture *) scr->icon_back_texture);
+    scr->icon_back_texture = wTextureMakeSolid(scr, &((*texture)->any.color));*/
+
+  /* Free the texture as nobody else will use it, nor refer to it.  */
+  if (!entry->addr)
+    wTextureDestroy(scr, *texture);
+
+  return (reset ? REFRESH_ICON_TILE : 0);
+}
+#endif
+
 static int setWinTitleFont(WScreen * scr, WDefaultEntry * entry, void *tdata, void *foo)
 {
 	WMFont *font = tdata;
