--- ./WindowMaker-0.95.7/src/framewin.c	2015-08-11 21:41:14.000000000 +0300
+++ ./WindowMaker/src/framewin.c	2016-11-25 00:09:13.797571299 +0200
@@ -1317,9 +1317,14 @@
 			XSetForeground(dpy, copy_gc, scr->black_pixel);
 			XDrawRectangle(dpy, button->window, copy_gc, 0, 0, button->width - 1, button->height - 1);
 		} else if (wPreferences.new_style == TS_OLD) {
+#ifdef NEXTSPACE
+			wDrawBevel(button->window, button->width, button->height, scr->widget_texture, WREL_RAISED);
+			XFillRectangle(dpy, button->window, copy_gc, 1, 1, button->width-3, button->height-3);
+#else
 			XFillRectangle(dpy, button->window, copy_gc, 0, 0, button->width, button->height);
 			XSetForeground(dpy, copy_gc, scr->black_pixel);
 			XDrawRectangle(dpy, button->window, copy_gc, 0, 0, button->width, button->height);
+#endif
 		} else {
 			XFillRectangle(dpy, button->window, copy_gc, 0, 0, button->width-3, button->height-3);
 			XSetForeground(dpy, copy_gc, scr->black_pixel);
@@ -1363,6 +1368,9 @@
 						  left, 0, width, image->height, x, y);
 			} else {
 				if (wPreferences.new_style == TS_OLD) {
+#ifdef NEXTSPACE
+					XSetClipOrigin(dpy, copy_gc, 2, 2);
+#endif
 					XSetForeground(dpy, copy_gc, scr->dark_pixel);
 					XFillRectangle(dpy, button->window, copy_gc, 0, 0,
 						       button->width, button->height);
