--- ./WindowMaker-0.95.7/src/placement.c	2015-08-11 21:41:14.000000000 +0300
+++ ./WindowMaker/src/placement.c	2018-06-24 23:33:20.535871415 +0300
@@ -66,7 +66,11 @@
 		   (((WIcon *) parent)->owner->frame->workspace == workspace
 		    || IS_OMNIPRESENT(((WIcon *) parent)->owner)
 		    || wPreferences.sticky_icons)
-		   && ((WIcon *) parent)->mapped) {
+		   && (((WIcon *) parent)->mapped
+#ifdef NEXTSPACE
+                       ||!((WIcon *) parent)->owner->screen_ptr->flags.icon_yard_mapped)
+#endif
+                   ) {
 
 		*retX = ((WIcon *) parent)->owner->icon_x;
 		*retY = ((WIcon *) parent)->owner->icon_y;
