--- ./WindowMaker-0.95.7/src/screen.h	2015-08-11 21:41:14.000000000 +0300
+++ ./WindowMaker/src/screen.h	2018-06-25 14:23:19.231149500 +0300
@@ -250,6 +250,9 @@
     WMScreen *wmscreen;		       /* for widget library */
 
     struct RImage *icon_tile;
+#ifdef NEXTSPACE
+    struct RImage *miniwindow_tile;
+#endif
     struct RImage *clip_tile;          /* tile with arrows to change workspace */
     struct RImage *drawer_tile;        /* tile for a drawer (tile + arrow) */
     Pixmap icon_tile_pixmap;		/* For app supplied icons */
@@ -310,6 +313,10 @@
         unsigned int doing_alt_tab:1;
         unsigned int jump_back_pending:1;
         unsigned int ignore_focus_events:1;
+#ifdef NEXTSPACE
+        unsigned int icon_yard_mapped:1;
+        unsigned int modifier_pressed:1;
+#endif
     } flags;
 } WScreen;
 
