--- ./wmaker-crm_next/WINGs/wcolorpanel.c	2015-07-29 16:02:49.000000000 +0300
+++ ./WM/WINGs/wcolorpanel.c	2015-08-01 22:32:01.606625969 +0300
@@ -397,7 +397,11 @@
 	panel->mode = WMWheelModeColorPanel;
 	panel->lastChanged = 0;
 	panel->slidersmode = WMRGBModeColorPanel;
+#ifdef NEXTSPACE
+	panel->configurationPath = wstrconcat(wusergnusteppath(), "/WindowMaker/Colors/");
+#else
 	panel->configurationPath = wstrconcat(wusergnusteppath(), "/Library/Colors/");
+#endif
 
 	/* Some General Purpose Widgets */
 	panel->colorWell = WMCreateColorWell(panel->win);
