--- ./WindowMaker-0.95.7/WINGs/wcolorpanel.c	2015-08-11 21:41:14.000000000 +0300
+++ ./WindowMaker/WINGs/wcolorpanel.c	2016-11-25 00:09:13.800571299 +0200
@@ -397,7 +397,11 @@
 	panel->mode = WMWheelModeColorPanel;
 	panel->lastChanged = 0;
 	panel->slidersmode = WMRGBModeColorPanel;
+#ifdef NEXTSPACE
+	panel->configurationPath = wstrconcat(wusergnusteppath(), "/WindowMaker/Colors/");
+#else
 	panel->configurationPath = wstrconcat(wusergnusteppath(), "/Library/Colors/");
+#endif
 
 	/* Some General Purpose Widgets */
 	panel->colorWell = WMCreateColorWell(panel->win);
