--- ./WindowMaker-0.95.7/WINGs/userdefaults.c	2015-08-11 21:41:14.000000000 +0300
+++ ./WindowMaker/WINGs/userdefaults.c	2016-11-24 23:11:41.533326672 +0200
@@ -40,7 +40,11 @@
 
 static void synchronizeUserDefaults(void *foo);
 
+#ifdef NEXTSPACE
+#define DEFAULTS_DIR "/Preferences/.WindowMaker"
+#else
 #define DEFAULTS_DIR "/Defaults"
+#endif
 #ifndef HAVE_INOTIFY
 /* Check defaults database for changes every this many milliseconds */
 /* XXX: this is shared with src/ stuff, put it in some common header */
