--- ./WindowMaker-0.95.7/WINGs/wcolor.c	2015-08-11 21:41:14.000000000 +0300
+++ ./WindowMaker/WINGs/wcolor.c	2017-11-15 09:55:42.523501721 +0200
@@ -245,7 +245,7 @@
 			WMReleaseColor(white);
 			WMReleaseColor(black);
 		} else {
-			color = WMCreateRGBColor(scr, 0xaeba, 0xaaaa, 0xaeba, True);
+			color = WMCreateRGBColor(scr, 0xaaaa, 0xaaaa, 0xaaaa, True);
 			if (!color->flags.exact)
 				wwarning(_("could not allocate %s color"), _("gray"));
 		}
@@ -283,7 +283,7 @@
 			WMReleaseColor(white);
 			WMReleaseColor(black);
 		} else {
-			color = WMCreateRGBColor(scr, 0x5144, 0x5555, 0x5144, True);
+			color = WMCreateRGBColor(scr, 0x5353, 0x5353, 0x5353, True);
 			if (!color->flags.exact)
 				wwarning(_("could not allocate %s color"), _("dark gray"));
 		}
