--- libdispatch-1.3.orig/dispatch/dispatch.h	2016-10-22 01:13:49.234193116 +0300
+++ libdispatch-1.3/dispatch/dispatch.h	2016-10-12 01:15:52.000000000 +0300
@@ -44,7 +44,15 @@
 #include <stdint.h>
 #include <stdbool.h>
 #include <stdarg.h>
-#include <unistd.h>
+
+// Fix RedHat problematic unistd.h
+#pragma push_macro("__block")
+#undef __block
+#define __block my__block
+#include_next <unistd.h>
+#pragma pop_macro("__block")
+//#include <unistd.h>
+
 #include <fcntl.h>
 
 #if defined(__linux__) && defined(__has_feature)
