diff -Nru libs-gui-gui-0_25_1.orig/Model/GNUmakefile libs-gui-gui-0_25_1/Model/GNUmakefile
--- libs-gui-gui-0_25_1.orig/Model/GNUmakefile	2017-04-07 00:03:24.000000000 +0300
+++ libs-gui-gui-0_25_1/Model/GNUmakefile	2019-03-29 15:43:57.705417000 +0200
@@ -66,9 +66,9 @@
 ADDITIONAL_LIB_DIRS += -L$(GNUSTEP_OBJ_DIR) 
 ifeq ($(GUI_LIB), gnu)
   ADDITIONAL_LIB_DIRS += -L../Source/$(GNUSTEP_OBJ_DIR)
-  ADDITIONAL_INCLUDE_DIRS = -I../Headers/Additions -I../Headers
+  ADDITIONAL_INCLUDE_DIRS += -I../Headers/Additions -I../Headers
 else
-  ADDITIONAL_INCLUDE_DIRS = -I../Headers/Additions
+  ADDITIONAL_INCLUDE_DIRS += -I../Headers/Additions
 endif
 
 ADDITIONAL_OBJC_FLAGS += $(BACKEND_DEFINE)
